# Exercise 4.7. Deriving New Variables

In [1]:
# Importing 
import pandas as pd
import numpy as np
import os

In [2]:
# Import dataframe 'ords_prods_merge'
df_ords_prods_merge = pd.read_pickle(r'C:\Users\ZenBook\Instacart Basket Analysis_30.09.2025\02 Data\Prepared Data\ords_prods_merge.pkl')

In [3]:
# subset - to work only with 10000 rows in the dataframe
df = df_ords_prods_merge[:1000000]

In [4]:
# check shape 'df_ords_prods_merge'
df.shape

(1000000, 16)

# If-Statements with User-Defined Functions

In [5]:
# defining the function 'price_label'

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [6]:
# apply the function to subset

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\ZenBook\AppData\Local\Temp\ipykernel_3572\2609210699.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [7]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652593
Low-range product    338063
High range             9344
Name: count, dtype: int64

In [8]:
# use the max function - within the “prices” column of the df dataframe
df['prices'].max()

24.5

# If-Statements with the loc() Function

In [9]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\ZenBook\AppData\Local\Temp\ipykernel_3572\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [10]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [11]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [12]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     652593
Low-range product     338063
High-range product      9344
Name: count, dtype: int64

get the same results when used user-defined function

In [13]:
# use loc() on entire dataframe 'ords_prods_merge'
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [14]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [15]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [16]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21861997
Low-range product     10126366
High-range product      417678
Name: count, dtype: int64

# If-Statements with For-Loops

how busy each day of the week 

on which day most orders take place? - frequency of the “orders_dow” column

In [17]:
df_ords_prods_merge['order_dow'].value_counts(dropna = False)

order_dow
0    6204404
1    5660456
6    4496635
2    4213986
5    4205906
3    3840701
4    3783953
Name: count, dtype: int64

0 - Saturday is the busiest day (the day when the most orders take place)
4 - Wednesday is the least busy day

In [18]:
# create a new column “busiest day” using for-loop
result = []

for value in df_ords_prods_merge["order_dow"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

This for loop will check every single value under the 'orders_dow'. Then, it will apply the conditions above. After that, it will push the condition to the 'result = []' list 

In [19]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

create a new column within df_ords_prods_merge dataframe, make it equal to result

In [20]:
df_ords_prods_merge['busiest_day'] = result

In [21]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22417684
Busiest day        6204404
Least busy         3783953
Name: count, dtype: int64

# ANSWERS TO TASK

# Suppose your clients have changed their minds about the labels you created in your “busiest_day” column.

“Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days.

In [22]:
# create a new column “busiest days” using for-loop
results = []

for value in df_ords_prods_merge["order_dow"]:
  if value in (0, 1):
    results.append("Busiest days")
  elif value in (3, 4):
    results.append("Slowest days")
  else:
    results.append("Regular days")

Saturday (0) and Sunday (1) are the busiest days,
Wednesday (3) and Thursday (4) are the Slowest days,

On every loop, the 'if value in (0, 1)' will check if the day is either Saturday (0) OR Sunday (1); if it is, then it will add 'Busiest days' to 'results' list.

If it's not Saturday OR Sunday, the second line (elif value in (3, 4:) will execute and check whether the day is either Wednesday (3) or Thursday (4). Then 'Slowest days' will be printed to the 'results' list.

If none of the two previous conditions is met, then 'else' will be executed, meaning that 'Regular days' will be pushed to the results list.

In [23]:
results

['Regular days',
 'Regular days',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Regular days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Slowest days',
 'Slowest days',
 'Regular days',
 'Slowest days',
 'Regular days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regular days',
 'Busiest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Busiest days

In [24]:
# creating the 'busiest_days' column
df_ords_prods_merge['busiest_days'] = results

In [25]:
df_ords_prods_merge.shape

(32406041, 19)

In [26]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   3139998      138    prior            28          6                 11   
1   1977647      138    prior            30          6                 17   
2    389851      709    prior             2          0                 21   
3    652770      764    prior             1          3                 13   
4   1813452      764    prior             3          4                 17   

   days_since_prior_order  first_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   
3                     NaN         True                 10          0   both   
4                     9.0        False                 11          1   both   

     price_range_loc     busiest_day  busiest_days  
0  Mid-range product  Regularly busy  Regular days  
1  Mid-range product  Regularly busy  Regular days  
2  Mid-range product     Busiest day  Busiest days  
3  Mid-range product  Regularly busy  Slowest days  
4  Mid-range product      Least busy  Slowest days

In [27]:
# check the values of this new column for accuracy.
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regular days    12916527
Busiest days    11864860
Slowest days     7624654
Name: count, dtype: int64

The sum of the above values is 32406041.

# Identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [28]:
# receive list from the busiest hour to the least busy hour
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761862
11    2736242
14    2689251
15    2662246
13    2661057
12    2618613
16    2535313
9     2454273
17    2087721
8     1718182
18    1636566
19    1258352
20     976182
7      891082
21     795658
22     634247
23     402335
6      290500
0      218786
1      115706
5       87964
2       69377
4       53245
3       51281
Name: count, dtype: int64

Most orders happened on 9,10,11,12,13,14,15,16 hour of the day; Fewest orders - 0,1,2,3,4,5,6,23 hour of the day; Average orders in the middle - 7, 8, 17-22 hour of the day.

In [29]:
results_hours = []

for value in df_ords_prods_merge["order_hour_of_day"]:
    if value in (9, 10, 11, 12, 13, 14, 15, 16):
        results_hours.append("Most orders")
    elif value in (7, 8, 17, 18, 19, 20, 21, 22):
        results_hours.append("Average orders")
    else:
        results_hours.append("Fewest orders")

In [30]:
# creating the column 'busiest_period_of_day'
df_ords_prods_merge['busiest_period_of_day'] = results_hours

In [31]:
# the frequency in the 'busiest_period_of_day'
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118857
Average orders     9997990
Fewest orders      1289194
Name: count, dtype: int64

The sum of the above values is 222406041

In [32]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   3139998      138    prior            28          6                 11   
1   1977647      138    prior            30          6                 17   
2    389851      709    prior             2          0                 21   
3    652770      764    prior             1          3                 13   
4   1813452      764    prior             3          4                 17   

   days_since_prior_order  first_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   
3                     NaN         True                 10          0   both   
4                     9.0        False                 11          1   both   

     price_range_loc     busiest_day  busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regular days           Most orders  
1  Mid-range product  Regularly busy  Regular days        Average orders  
2  Mid-range product     Busiest day  Busiest days        Average orders  
3  Mid-range product  Regularly busy  Slowest days           Most orders  
4  Mid-range product      Least busy  Slowest days        Average orders

# Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [33]:
# creating path
path = r'C:\Users\ZenBook\Instacart Basket Analysis_30.09.2025'

In [34]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_nvariabl.pkl'))